### PR TITLE
Test: Add test to verify requests are not scheduled if limit is exceeded

### PR DESCRIPTION
**Pull request by sseshasa**
Add a test to verify that when a client exceeds its limit, the
scheduler does not schedule the requests from that client but instead
responds with a future request type until the limit is restored.

The motivation for adding this test is that this scenario is currently not
handled within the mClockScheduler::dequeue() function i.e. handle
future request type response. The code to handle this case is currently
being put in place and tested as part of another PR.

### CODE REVIEW

**Review by athanatos — APPROVED** [review]

File: test/test_dmclock_server.cc [hunk 1]
```diff
@@ -1359,5 +1359,110 @@ namespace crimson {
...
+      ClientId client1 = 52;
+      ClientId client2 = 8;
+
+      // Create client1 with high limit.
+      // Create client2 with low limit and with lower weight than client1
+      dmc::ClientInfo info1(1.0, 2.0, 100.0);
+      dmc::ClientInfo info2(1.0, 1.0, 2.0);
+
+      auto client_info_f = [&] (ClientId c) -> const dmc::ClientInfo* {
+        if (client1 == c) return &info1;
```
Comment (ivancich) [inline test/test_dmclock_server.cc:1377]: Please use ceph coding style as documented at https://github.com/ceph/ceph/blob/master/CodingStyle .

File: test/test_dmclock_server.cc [hunk 2]
```diff
@@ -1359,5 +1359,110 @@ namespace crimson {
...
+      for (int i = 0; i< 2; ++i) {
+        Queue::PullReq pr = pq->pull_request();
+        EXPECT_EQ(Queue::NextReqType::returning, pr.type);
+        auto& retn = boost::get<Queue::PullReq::Retn>(pr.data);
+
+        if (client1 == retn.client) {
+          ++c1_count;
+        } else if (client2 == retn.client) {
+          ++c2_count;
+        } else ADD_FAILURE() << "got request from neither of two clients";
```
Comment (ivancich) [inline test/test_dmclock_server.cc:1415]: Please use ceph coding style as documented at https://github.com/ceph/ceph/blob/master/CodingStyle .

File: test/test_dmclock_server.cc [hunk 3]
```diff
@@ -1359,5 +1359,110 @@ namespace crimson {
...
+      for (int i = 0; i < 50; ++i) {
+        Queue::PullReq pr = pq->pull_request();
+        EXPECT_EQ(Queue::NextReqType::returning, pr.type);
+        auto& retn = boost::get<Queue::PullReq::Retn>(pr.data);
+
+        if (client1 == retn.client) {
+          ++c1_count;
+        } else if (client2 == retn.client) {
+          ++c2_count;
+        } else ADD_FAILURE() << "got request from neither of two clients";
```
Comment (ivancich) [inline test/test_dmclock_server.cc:1440]: Please use ceph coding style as documented at https://github.com/ceph/ceph/blob/master/CodingStyle .

File: test/test_dmclock_server.cc [hunk 4]
```diff
@@ -1359,5 +1359,110 @@ namespace crimson {
...
+      }
+
+      EXPECT_EQ(2u, pq->client_count());
+      EXPECT_EQ(100u, pq->request_count());
+      int c1_count = 0;
+      int c2_count = 0;
+
+      // Pull couple of requests, should come from reservation queue.
+      // One request each from client1 and client2 should be pulled.
+      for (int i = 0; i< 2; ++i) {
```
Comment (ivancich) [inline test/test_dmclock_server.cc:1406]: missing space

**Review by ivancich — CHANGES_REQUESTED** [review]
Looks good overall. Please use ceph coding style as documented at https://github.com/ceph/ceph/blob/master/CodingStyle . Otherwise, I approve.

File: test/test_dmclock_server.cc (hunk 1 above)
Comment (sseshasa) [reply test/test_dmclock_server.cc]: Done.

File: test/test_dmclock_server.cc (hunk 2 above)
Comment (sseshasa) [reply test/test_dmclock_server.cc]: Done.

File: test/test_dmclock_server.cc (hunk 3 above)
Comment (sseshasa) [reply test/test_dmclock_server.cc]: Done.

File: test/test_dmclock_server.cc (hunk 4 above)
Comment (sseshasa) [reply test/test_dmclock_server.cc]: Fixed.

**Comment by sseshasa** [comment]
> Looks good overall. Please use ceph coding style as documented at https://github.com/ceph/ceph/blob/master/CodingStyle . Otherwise, I approve.

Thanks! There are other tests in the file where similar coding style needs to be adopted. That could be taken up as part of another PR.

**Comment by ivancich** [comment]
> > Looks good overall. Please use ceph coding style as documented at https://github.com/ceph/ceph/blob/master/CodingStyle . Otherwise, I approve.
> 
> Thanks! There are other tests in the file where similar coding style needs to be adopted. That could be taken up as part of another PR.

Agreed, a clean-up PR is in order. And I'm probably responsible for some of the code that needs the clean-up.

**Review by ivancich — APPROVED** [review]

**Pull request merged**